crates/mpc-protocols/src/mpc_service.rs

```diff
@@ -0,0 +1,115 @@
+use crate::mpc_events::{CreatedProofMPCEvent, MPCEvent};
+use pera_types::base_types::ObjectID;
+use pera_types::event::Event;
+use std::cell::RefCell;
+use std::collections::{HashMap, VecDeque};
+use tokio::sync::{mpsc, Mutex, RwLock};
+
+const MAX_ACTIVE_MPC_INSTANCES: usize = 100;
+
+pub enum MPCInput {
+    InitEvent(CreatedProofMPCEvent),
+    Message,
+    OutputEvent,
+}
+
+struct MPCInstance {
+    status: MPCStatus,
+    /// The channel to send message to this instance's message handler thread when this instance is active
+    messages_handler_sender: Option<mpsc::Sender<MPCInput>>,
+    pending_messages: Vec<MPCInput>,
+}
+
+/// Possible status of an MPC session
+/// - Active: The session is currently running, new messages will be forwarded to the session
+/// - Pending: The session is waiting for a slot to become active, when the init received there were already `MAX_ACTIVE_MPC_INSTANCES` active sessions
+/// - Finished: The session has finished, no more messages will be forwarded. The session will be removed from the service after a timeout.
+/// We want to keep it for some time so leftover messages related to it won't be treated as malicious.
+#[derive(Clone, Copy)]
+enum MPCStatus {
+    Active,
+    Pending,
+    Finished,
+}
```

Comment:
I'm not a fan of comments either but Zeev asked me to over-document the code so it will be easier for him to grasp it when he comes back. (He's on a holiday rn)

Reply:
Re "Pending to AwaitingInput" - I prefer pending

Reply:
I see that `AwaitingInput` is indeed not a good name. Will start a new comment to address my new feedback.